calour/heatmap/plotgui.py

```diff
@@ -328,12 +328,20 @@ def key_press_callback(self, event):
                 xlim_lower,
                 xlim_lower + (xlim_upper - xlim_lower) * self.zoom_scale)
         elif event.key == 'down':
+            if ylim_lower - y_offset > self.exp.data.shape[1] - 0.5:
+                y_offset = ylim_lower + 0.5 - self.exp.data.shape[1]
             ax.set_ylim(ylim_lower - y_offset, ylim_upper - y_offset)
         elif event.key == 'up':
+            if ylim_upper + y_offset < -0.5:
+                y_offset = -0.5 - ylim_upper
             ax.set_ylim(ylim_lower + y_offset, ylim_upper + y_offset)
         elif event.key == 'left':
+            if xlim_lower - x_offset < -0.5:
+                x_offset = xlim_lower + 0.5
             ax.set_xlim(xlim_lower - x_offset, xlim_upper - x_offset)
         elif event.key == 'right':
+            if xlim_upper + x_offset > self.exp.data.shape[0] - 0.5:
```

Comment:
minor, but use `right_end = self.exp.data.shape[0] - 0.5` to reuse the value below?

same for `down` key